README.md
# css-color-converter [![Build Status][ci-img]][ci]

Converts CSS colors from one representation to another 

[ci-img]:  https://travis-ci.org/andyjansson/css-color-converter.svg
[ci]:      https://travis-ci.org/andyjansson/css-color-converter

## Installation

```js
npm install css-color-converter
```

## Usage

**Example usage**: 
```js
import { fromString } from 'css-color-converter';

fromString('rgb(255, 255, 255)').toHslString(); // hsl(0, 0%, 100%)
fromString('rgba(255, 255, 255, 0.5)').toHslString(); // hsla(0, 0%, 100%, 0.5)
fromString('blue').toRgbString(); // rgb(0, 0, 255)
fromString('red').toHexString(); // #ff0000
```
## Functions

### `fromString(str)`
| parameter | type   | description                                    |
| --------- | ------ | ---------------------------------------------- |
| `str`     | string | Supports named colors, hex, rgb/rgba, hsl/hsla |

**Returns** `instance`

**Raises** `Error`

### `fromRgb([r, g, b])`

| parameter | type   | description   |
| --------- | ------ | ------------- |
| `r`       | int    | red (0-255)   |
| `g`       | int    | green (0-255) |
| `b`       | int    | blue (0-255)  |

**Returns** `instance`

### `fromRgba([r, g, b, a])`

| parameter | type   | description   |
| --------- | ------ | ------------- |
| `r`       | int    | red (0-255)   |
| `g`       | int    | green (0-255) |
| `b`       | int    | blue (0-255)  |
| `a`       | float  | alpha (0-1)   |

**Returns** `instance`

### `fromHsl([h, s, l])`

| parameter | type   | description        |
| --------- | ------ | ------------------ |
| `h`       | int    | hue (0-360)        |
| `s`       | int    | saturation (0-100) |
| `l`       | int    | luminosity (0-100) |

**Returns** `instance`

### `fromHsla([h, s, l, a])`

| parameter | type   | description        |
| --------- | ------ | ------------------ |
| `h`       | int    | hue (0-360)        |
| `s`       | int    | saturation (0-100) |
| `l`       | int    | luminosity (0-100) |
| `a`       | float  | alpha (0-1)        |

**Returns** `instance`

## Methods

### `toRgbString()`

**Returns** `rgb()` or `rgba()`, depending on the alpha.

### `toHslString()`

**Returns** `hsl()` or `hsla()`, depending on the alpha.

### `toHexString()`

**Returns** 6-digit or 8-digit `hex`, depending on the alpha.

### `toRgbaArray()`

**Returns** `[r, g, b, a]` array.

CHANGELOG.md
# 2.0.0 - Unreleased
* Rewrite as ES module. Some breaking changes to the API surface has been made as a result.
* Better adhere to CSS specification (#4).
* Add support for CSS Color Module Level 4 syntax.
* Update dependencies.
* Add eslint.

# 1.1.1 - 2020-03-20
* Fix alpha 0 incorrectly defaulting to 1 (#2). Thanks to jedwards1211.

# 1.1.0 - 2016-08-13
* Add `toHslaArray` method.

# 1.0.2 - 2015-05-27
* Fix typo causing `rgba()` to return incorrect results.

# 1.0.1 - 2015-05-27
* Fix `toHexString()` not padding values.

# 1.0.0 - 2015-05-27
* Initial release.


test/test.js
var color = require(".."),
	assert = require("assert");

assert.deepEqual(color("blue").toRgbaArray(), [0, 0, 255, 1]);
assert.deepEqual(color("blue").toHslaArray(), [240, 100, 50, 1]);
assert.deepEqual(color("blue").toHexString(), '#0000ff');
assert.deepEqual(color("rgb(255, 255, 255)").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("rgb(100%, 100%, 100%)").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("rgb(100%, 100%, 100%)").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("rgba(100%, 100%, 100%, .5)").toHexString(), '#ffffff7f');
assert.deepEqual(color("rgba(255, 255, 255, .5)").toRgbaArray(), [255, 255, 255, .5]);
assert.deepEqual(color("rgba(255, 255, 255, .5)").toHslaArray(), [0, 0, 100, .5]);
assert.equal(color().fromRgb([255, 255, 255]).toRgbString(), 'rgb(255, 255, 255)');
assert.equal(color().fromRgba([255, 255, 255, 1]).toRgbString(), 'rgb(255, 255, 255)');
assert.equal(color().fromRgba([255, 255, 255, .5]).toRgbString(), 'rgba(255, 255, 255, 0.5)');
assert.equal(color().fromRgba([255, 255, 255, 1]).toHslString(), 'hsl(0, 0%, 100%)');
assert.equal(color().fromRgba([255, 255, 255, .5]).toHslString(), 'hsla(0, 0%, 100%, 0.5)');
assert.equal(color().fromRgba([255, 255, 255, 1]).toHexString(), '#ffffff');
assert.equal(color().fromRgba([255, 255, 255, .5]).toHexString(), '#ffffff7f');
assert.deepEqual(color("hsl(0, 0%, 100%)").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("hsl(0, 0%, 100%)").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("hsla(0, 0%, 100%, .5)").toHslaArray(), [0, 0, 100, .5]);
assert.deepEqual(color("hsla(0, 0%, 100%, .5)").toRgbaArray(), [255, 255, 255, .5]);
assert.equal(color().fromHsl([0, 0, 100]).toHslString(), 'hsl(0, 0%, 100%)');
assert.equal(color().fromHsla([0, 0, 100, 1]).toRgbString(), 'rgb(255, 255, 255)');
assert.equal(color().fromHsla([0, 0, 100, .5]).toRgbString(), 'rgba(255, 255, 255, 0.5)');
assert.equal(color().fromHsla([0, 0, 100, 1]).toHslString(), 'hsl(0, 0%, 100%)');
assert.equal(color().fromHsla([0, 0, 100, .5]).toHslString(), 'hsla(0, 0%, 100%, 0.5)');
assert.equal(color().fromHsla([0, 0, 100, 1]).toHexString(), '#ffffff');
assert.equal(color().fromHsla([0, 0, 100, .5]).toHexString(), '#ffffff7f');
assert.deepEqual(color("#ffffff").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("#ffffff").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("#ffffffff").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("#ffffffff").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("#fff").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("#fff").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("#ffff").toRgbaArray(), [255, 255, 255, 1]);
assert.deepEqual(color("#ffff").toHslaArray(), [0, 0, 100, 1]);
assert.deepEqual(color("#ffffff7f").toRgbaArray(), [255, 255, 255, .5]);
assert.deepEqual(color("#ffffff7f").toHslaArray(), [0, 0, 100, .5]);
assert.deepEqual(color("#ffffff").toHexString(), '#ffffff');
assert.equal(color("#ffffff").toRgbString(), 'rgb(255, 255, 255)');
assert.equal(color("#ffffff7f").toRgbString(), 'rgba(255, 255, 255, 0.5)');
assert.equal(color("#ffffff").toHslString(), 'hsl(0, 0%, 100%)');
assert.equal(color("#ffffff7f").toHslString(), 'hsla(0, 0%, 100%, 0.5)');
assert.equal(color("#ffffff").toHexString(), '#ffffff');
assert.equal(color("#ffffff7f").toHexString(), '#ffffff7f');
assert.deepEqual(color([0, 0, 0, 0]).toRgbaArray(), [0, 0, 0, 0]);
assert.deepEqual(color('rgba(0, 0, 0, 0)').toRgbaArray(), [0, 0, 0, 0]);
assert.deepEqual(color([3, 2, 1, 0]).toRgbaArray(), [3, 2, 1, 0]);
assert.deepEqual(color('rgba(3, 2, 1, 0)').toRgbaArray(), [3, 2, 1, 0]);


src/__tests__/test.mjs
import { deepStrictEqual, strictEqual, throws } from 'assert';
import {
  fromString, fromRgb, fromRgba, fromHsl, fromHsla,
} from '../index.mjs';

deepStrictEqual(fromString('blue').toRgbaArray(), [0, 0, 255, 1]);
deepStrictEqual(fromString('blue').toHslaArray(), [240, 100, 50, 1]);
deepStrictEqual(fromString('blue').toHexString(), '#0000ff');
deepStrictEqual(fromString('rgb(255, 255, 255)').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('rgb(100%, 100%, 100%)').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('rgb(100%, 100%, 100%)').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('rgba(100%, 100%, 100%, .5)').toHexString(), '#ffffff7f');
deepStrictEqual(fromString('rgba(255, 255, 255, .5)').toRgbaArray(), [255, 255, 255, 0.5]);
deepStrictEqual(fromString('rgba(255, 255, 255, .5)').toHslaArray(), [0, 0, 100, 0.5]);
deepStrictEqual(fromString('rgb(255 255 255)').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('rgb(100% 100% 100%)').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('rgb(100% 100% 100%)').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('rgba(100% 100% 100% / .5)').toHexString(), '#ffffff7f');
deepStrictEqual(fromString('rgba(255 255 255 / .5)').toRgbaArray(), [255, 255, 255, 0.5]);
deepStrictEqual(fromString('rgba(255 255 255 / .5)').toHslaArray(), [0, 0, 100, 0.5]);
deepStrictEqual(fromString('rgba(0, 0, 0, 0)').toRgbaArray(), [0, 0, 0, 0]);
deepStrictEqual(fromString('rgba(3, 2, 1, 0)').toRgbaArray(), [3, 2, 1, 0]);
deepStrictEqual(fromString('hsl(0, 0%, 100%)').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('hsl(0, 0%, 100%)').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('hsla(0, 0%, 100%, .5)').toHslaArray(), [0, 0, 100, 0.5]);
deepStrictEqual(fromString('hsla(0, 0%, 100%, .5)').toRgbaArray(), [255, 255, 255, 0.5]);
deepStrictEqual(fromString('#ffffff').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('#ffffff').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('#ffffffff').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('#ffffffff').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('#fff').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('#fff').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('#ffff').toRgbaArray(), [255, 255, 255, 1]);
deepStrictEqual(fromString('#ffff').toHslaArray(), [0, 0, 100, 1]);
deepStrictEqual(fromString('#ffffff7f').toRgbaArray(), [255, 255, 255, 0.5]);
deepStrictEqual(fromString('#ffffff7f').toHslaArray(), [0, 0, 100, 0.5]);
deepStrictEqual(fromString('#ffffff').toHexString(), '#ffffff');
strictEqual(fromString('#ffffff').toRgbString(), 'rgb(255, 255, 255)');
strictEqual(fromString('#ffffff7f').toRgbString(), 'rgba(255, 255, 255, 0.5)');
strictEqual(fromString('#ffffff00').toRgbString(), 'rgba(255, 255, 255, 0)');
strictEqual(fromString('#fff0').toRgbString(), 'rgba(255, 255, 255, 0)');
strictEqual(fromString('#ffffff').toHslString(), 'hsl(0, 0%, 100%)');
strictEqual(fromString('#ffffff7f').toHslString(), 'hsla(0, 0%, 100%, 0.5)');
strictEqual(fromString('#ffffff').toHexString(), '#ffffff');
strictEqual(fromString('#ffffff7f').toHexString(), '#ffffff7f');
throws(() => fromString('nonsense'));
strictEqual(fromRgb([255, 255, 255]).toRgbString(), 'rgb(255, 255, 255)');
strictEqual(fromRgba([255, 255, 255, 1]).toRgbString(), 'rgb(255, 255, 255)');
strictEqual(fromRgba([255, 255, 255, 0.5]).toRgbString(), 'rgba(255, 255, 255, 0.5)');
strictEqual(fromRgba([255, 255, 255, 1]).toHslString(), 'hsl(0, 0%, 100%)');
strictEqual(fromRgba([255, 255, 255, 0.5]).toHslString(), 'hsla(0, 0%, 100%, 0.5)');
strictEqual(fromRgba([255, 255, 255, 1]).toHexString(), '#ffffff');
strictEqual(fromRgba([255, 255, 255, 0.5]).toHexString(), '#ffffff7f');
deepStrictEqual(fromRgba([0, 0, 0, 0]).toRgbaArray(), [0, 0, 0, 0]);
deepStrictEqual(fromRgba([3, 2, 1, 0]).toRgbaArray(), [3, 2, 1, 0]);
strictEqual(fromHsl([0, 0, 100]).toHslString(), 'hsl(0, 0%, 100%)');
strictEqual(fromHsla([0, 0, 100, 1]).toRgbString(), 'rgb(255, 255, 255)');
strictEqual(fromHsla([0, 0, 100, 0.5]).toRgbString(), 'rgba(255, 255, 255, 0.5)');
strictEqual(fromHsla([0, 0, 100, 1]).toHslString(), 'hsl(0, 0%, 100%)');
strictEqual(fromHsla([0, 0, 100, 0.5]).toHslString(), 'hsla(0, 0%, 100%, 0.5)');
strictEqual(fromHsla([0, 0, 100, 1]).toHexString(), '#ffffff');
strictEqual(fromHsla([0, 0, 100, 0.5]).toHexString(), '#ffffff7f');


src/index.mjs
import colorNames from 'color-name';
import unitConverter from 'css-unit-converter';

const hex = /^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?$/;
const shortHex = /^#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])?$/;
const rgb = /^rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)(?:\s*,\s*(0|1|0?\.\d+|\d+%))?\s*\)$/;
const rgbfn = /^rgba?\(\s*(\d+)\s+(\d+)\s+(\d+)(?:\s*\/\s*(0|1|0?\.\d+|\d+%))?\s*\)$/;
const rgbperc = /^rgba?\(\s*(\d+%)\s*,\s*(\d+%)\s*,\s*(\d+%)(?:\s*,\s*(0|1|0?\.\d+|\d+%))?\s*\)$/;
const rgbpercfn = /^rgba?\(\s*(\d+%)\s+(\d+%)\s+(\d+%)(?:\s*\/\s*(0|1|0?\.\d+|\d+%))?\s*\)$/;
const hsl = /^hsla?\(\s*(\d+)(deg|rad|grad|turn)?\s*,\s*(\d+)%\s*,\s*(\d+)%(?:\s*,\s*(0|1|0?\.\d+|\d+%))?\s*\)$/;

function contains(haystack, needle) {
  return haystack.indexOf(needle) > -1;
}

function rgbToHsl(r, g, b) {
  const rprim = r / 255;
  const gprim = g / 255;
  const bprim = b / 255;
  const cmax = Math.max(rprim, gprim, bprim);
  const cmin = Math.min(rprim, gprim, bprim);
  const delta = cmax - cmin;
  const l = (cmax + cmin) / 2;

  if (delta === 0) {
    return [0, 0, l * 100];
  }
  const s = delta / (1 - Math.abs(2 * l - 1));
  const h = (() => {
    switch (cmax) {
      case rprim: {
        return ((gprim - bprim) / delta) % 6;
      }
      case gprim: {
        return (bprim - rprim) / delta + 2;
      }
      default: {
        return (rprim - gprim) / delta + 4;
      }
    }
  })();

  return [
    h * 60,
    s * 100,
    l * 100,
  ];
}

function hslToRgb(h, s, l) {
  const hprim = h / 60;
  const sprim = s / 100;
  const lprim = l / 100;
  const c = (1 - Math.abs(2 * lprim - 1)) * sprim;
  const x = c * (1 - Math.abs((hprim % 2) - 1));
  const m = lprim - c / 2;
  const [rprim, gprim, bprim] = (() => {
    if (hprim < 1) return [c, x, 0];
    if (hprim < 2) return [x, c, 0];
    if (hprim < 3) return [0, c, x];
    if (hprim < 4) return [0, x, c];
    if (hprim < 5) return [x, 0, c];
    return [c, 0, x];
  })();

  return [
    (rprim + m) * 255,
    (gprim + m) * 255,
    (bprim + m) * 255,
  ];
}

class Color {
  constructor([r, g, b, a]) {
    this.values = [
      Math.max(Math.min(parseInt(r, 10), 255), 0),
      Math.max(Math.min(parseInt(g, 10), 255), 0),
      Math.max(Math.min(parseInt(b, 10), 255), 0),
      a == null ? 1 : Math.max(Math.min(parseFloat(a), 255), 0),
    ];
  }

  toRgbString() {
    const [r, g, b, a] = this.values;
    if (a === 1) {
      return `rgb(${r}, ${g}, ${b})`;
    }
    return `rgba(${r}, ${g}, ${b}, ${a})`;
  }

  toHslString() {
    const [h, s, l, a] = this.toHslaArray();
    if (a === 1) {
      return `hsl(${h}, ${s}%, ${l}%)`;
    }
    return `hsla(${h}, ${s}%, ${l}%, ${a})`;
  }

  toHexString() {
    let [r, g, b, a] = this.values;
    r = Number(r).toString(16).padStart(2, '0');
    g = Number(g).toString(16).padStart(2, '0');
    b = Number(b).toString(16).padStart(2, '0');
    a = a < 1 ? parseInt(a * 255, 10).toString(16).padStart(2, '0') : '';
    return `#${r}${g}${b}${a}`;
  }

  toRgbaArray() {
    return this.values;
  }

  toHslaArray() {
    const [r, g, b, a] = this.values;
    const [h, s, l] = rgbToHsl(r, g, b);
    return [h, s, l, a];
  }
}

export function fromRgba([r, g, b, a]) {
  return new Color([r, g, b, a]);
}

export function fromRgb([r, g, b]) {
  return fromRgba([r, g, b, 1]);
}

export function fromHsla([h, s, l, a]) {
  const [r, g, b] = hslToRgb(h, s, l);
  return fromRgba([r, g, b, a]);
}

export function fromHsl([h, s, l]) {
  return fromHsla([h, s, l, 1]);
}

function fromHexString(str) {
  let [, r, g, b, a] = hex.exec(str) || shortHex.exec(str);
  r = parseInt(r.length < 2 ? r.repeat(2) : r, 16);
  g = parseInt(g.length < 2 ? g.repeat(2) : g, 16);
  b = parseInt(b.length < 2 ? b.repeat(2) : b, 16);
  a = (a && (parseInt(a.length < 2 ? a.repeat(2) : a, 16) / 255).toPrecision(1)) || 1;
  return fromRgba([r, g, b, a]);
}

function fromRgbString(str) {
  let [, r, g, b, a] = rgb.exec(str) || rgbperc.exec(str) || rgbfn.exec(str) || rgbpercfn.exec(str);
  r = contains(r, '%') ? (parseInt(r, 10) * 255) / 100 : parseInt(r, 10);
  g = contains(g, '%') ? (parseInt(g, 10) * 255) / 100 : parseInt(g, 10);
  b = contains(b, '%') > 0 ? (parseInt(b, 10) * 255) / 100 : parseInt(b, 10);
  a = a === undefined ? 1 : parseFloat(a) / (contains(a, '%') ? 100 : 1);
  return fromRgba([r, g, b, a]);
}

function fromHslString(str) {
  let [, h, unit, s, l, a] = hsl.exec(str);
  unit = unit || 'deg';
  h = unitConverter(parseFloat(h), unit, 'deg');
  s = parseFloat(s);
  l = parseFloat(l);
  a = a === undefined ? 1 : parseFloat(a) / (contains(a, '%') ? 100 : 1);
  return fromHsla([h, s, l, a]);
}

export function fromString(str) {
  if (colorNames[str]) {
    return fromRgb(colorNames[str]);
  }

  if (hex.test(str) || shortHex.test(str)) {
    return fromHexString(str);
  }

  if (rgb.test(str) || rgbperc.test(str) || rgbfn.test(str) || rgbpercfn.test(str)) {
    return fromRgbString(str);
  }

  if (hsl.test(str)) {
    return fromHslString(str);
  }

  throw Error(`Cannot parse string '${str}'`);
}

export default {
  fromString,
  fromRgb,
  fromRgba,
  fromHsl,
  fromHsla,
};


index.js
var colorNames = require('color-name'),
	converter = require('color-convert')();

var Color = function (obj) {
	if (!(this instanceof Color)) return new Color(obj);

	if (typeof obj == "string") {
		var hex = /^#([0-9a-fA-F]{6}(?:[0-9a-fA-F]{2})?)$/,
			shortHex = /^#([0-9a-fA-F]{3}[0-9a-fA-F]?)$/,
			rgb = /^rgb\(\s*([0-9]+%?)\s*,\s*([0-9]+%?)\s*,\s*([0-9]+%?)\s*\)$/,
			rgba = /^rgba\(\s*([0-9]+%?)\s*,\s*([0-9]+%?)\s*,\s*([0-9]+%?)\s*,\s*([0-1]|0?\.[0-9]+)\s*\)$/,
			hsl = /^hsl\(\s*([0-9]+)\s*,\s*([0-9]+)%\s*,\s*([0-9]+)%\s*\)$/,
			hsla = /^hsla\(\s*([0-9]+)\s*,\s*([0-9]+)%\s*,\s*([0-9]+)%\s*,\s*([0-1]|0?\.[0-9]+)\s*\)$/,
			match;

		if (match = colorNames[obj]) {
			return this.fromRgb(match);
		}
		else if (match = hex.exec(obj)) {
			var rgb = [];
			for (var i = 0; i < match[1].length / 2; i++) {
				rgb[i] = parseInt(match[1].slice(i * 2, i * 2 + 2), 16);
			}
			if (rgb.length == 4) {
				rgb[3] = (rgb[3] / 255).toPrecision(1);
				return this.fromRgba(rgb);
			}
			return this.fromRgb(rgb);
		}
		else if (match = shortHex.exec(obj)) {
			var rgb = [];
			for (var i = 0; i < match[1].length; i++) {
				rgb[i] = parseInt(match[1][i] + match[1][i], 16);
			}
			if (rgb.length == 4) {
				rgb[3] = (rgb[3] / 255).toPrecision(1);
				return this.fromRgba(rgb);
			}
			return this.fromRgb(rgb);
		}
		else if (match = rgb.exec(obj)) {
			var rgb = [
				parseInt(match[1]),
				parseInt(match[2]),
				parseInt(match[3])
			];
			if (match[1].indexOf("%") > -1) rgb[0] = rgb[0]*255/100;
			if (match[2].indexOf("%") > -1) rgb[1] = rgb[1]*255/100;
			if (match[3].indexOf("%") > -1) rgb[2] = rgb[2]*255/100;

			return this.fromRgb(rgb);
		}
		else if (match = rgba.exec(obj)) {
			var rgba = [
				parseInt(match[1]),
				parseInt(match[2]),
				parseInt(match[3]),
				parseFloat(match[4])
			];
			if (match[1].indexOf("%") > -1) rgba[0] = rgba[0]*255/100;
			if (match[2].indexOf("%") > -1) rgba[1] = rgba[1]*255/100;
			if (match[3].indexOf("%") > -1) rgba[2] = rgba[2]*255/100;

			return this.fromRgba(rgba);
		}
		else if (match = hsl.exec(obj)) {
			var hsl = [
				parseInt(match[1]),
				parseInt(match[2]),
				parseInt(match[3])
			];
			return this.fromHsl(hsl);
		}
		else if (match = hsla.exec(obj)) {
			var hsla = [
				parseInt(match[1]),
				parseInt(match[2]),
				parseInt(match[3]),
				parseFloat(match[4])
			];
			return this.fromHsla(hsla);
		}
	}

	if (Array.isArray(obj)) {

		if (obj.length < 3)
			throw new Error("Insufficient number of parameters");

		this.values = [
			Math.min(parseInt(obj[0]), 255),
			Math.min(parseInt(obj[1]), 255),
			Math.min(parseInt(obj[2]), 255),
			obj[3] == null ? 1 : parseFloat(obj[3])
		];
	}
};

Color.prototype = {
	fromRgba: function (vals) {
		return new Color(vals);
	},
	fromHsla: function (vals) {
		var rgb = converter.hsl(vals).rgb();
		rgb[3] = vals[3];
		return this.fromRgba(rgb);
	},
	fromRgb: function (vals) {
		vals[3] = 1;
		return this.fromRgba(vals);
	},
	fromHsl: function (vals) {
		vals[3] = 1;
		return this.fromHsla(vals);
	},

	toRgbString: function () {
		var rgb = this.values;
		if (rgb[3] === 1) return "rgb("+rgb[0]+", "+rgb[1]+", "+rgb[2]+")";
		return "rgba("+rgb[0]+", "+rgb[1]+", "+rgb[2]+", "+rgb[3]+")";
	},
	toHslString: function () {
		var hsl = converter.rgb(this.values).hsl();
		if (this.values[3] === 1) return "hsl("+hsl[0]+", "+hsl[1]+"%, "+hsl[2]+"%)";
		return "hsla("+hsl[0]+", "+hsl[1]+"%, "+hsl[2]+"%, "+this.values[3]+")";
	},
	toHexString: function () {
		var hex = "#";
		var r = this.values[0].toString(16);
		var g = this.values[1].toString(16);
		var b = this.values[2].toString(16);
		var a = parseInt(this.values[3] * 255).toString(16);
		if (r.length == 1) r = "0" + r;
		if (g.length == 1) g = "0" + g;
		if (b.length == 1) b = "0" + b;
		if (a.length == 1) a = "0" + a;
		hex += r + g + b;
		if (a !== "ff") hex += a;
		return hex;
	},
	toRgbaArray: function () {
		return this.values;
	},
	toHslaArray: function () {
		var hsla = converter.rgb(this.values).hsl();
		hsla[3] = this.values[3];
		return hsla;
	}
};


module.exports = Color;
